experiment/benchmark.py

```diff
@@ -83,32 +86,60 @@ def from_yaml(cls, benchmark_path: str) -> List:
-      benchmarks.append(
-          cls(truncated_id.lower(),
-              data['project'],
-              data['language'],
-              function.get('signature'),
-              function.get('name'),
-              function.get('return_type'),
-              function.get('params'),
-              function.get('exceptions', []),
-              function.get('is_jvm_static', False),
-              data['target_path'],
-              data.get('target_name'),
-              use_project_examples=use_project_examples,
-              cppify_headers=cppify_headers,
-              commit=commit,
-              use_context=use_context,
-              function_dict=function))
+
+    is_test_benchmark = data.get('is_test_benchmark', False)
+    test_files = data.get('test_files', [])
+    if is_test_benchmark:
+      for test_file in test_files:
+        max_len = os.pathconf('/', 'PC_NAME_MAX') - len('output-')
+        test_file_path = test_file.get('test_file_path')
+        truncated_id = f'{project_name}-{test_file_path.replace("/","_").replace(".","_")}'[:
+                                                                                            max_len]
+
+        benchmarks.append(
+            cls(
+                truncated_id.lower(),
+                data['project'],
+                data['language'],
+                '',
+                '',
+                '',
+                [],
+                [],
+                False,
+                data['target_path'],
+                data.get('target_name', ''),
+                is_test_benchmark=True,
+                test_file_path=test_file_path,
+            ))
+    else:
+      # function type benchmark
+      for function in functions:
+        # Long raw_function_names (particularly for c++ projects) may exceed
+        # filesystem limits on file path/name length when creating WorkDir.
+        max_len = os.pathconf('/', 'PC_NAME_MAX') - len('output-')
+        # Docker tag name cannot exceed 127 characters, and will be suffixed by
+        # '<sample-id>-experiment'.
+        docker_name_len = 127 - len('-03-experiment')
+        max_len = min(max_len, docker_name_len)
+        truncated_id = f'{project_name}-{function.get("name")}'[:max_len]
+        benchmarks.append(
+            cls(truncated_id.lower(),
+                data['project'],
+                data['language'],
+                function.get('signature'),
+                function.get('name'),
+                function.get('return_type'),
+                function.get('params'),
+                function.get('exceptions', []),
+                function.get('is_jvm_static', False),
+                data['target_path'],
+                data.get('target_name'),
+                use_project_examples=use_project_examples,
+                cppify_headers=cppify_headers,
+                commit=commit,
+                use_context=use_context,
+                function_dict=function))
```

Comment:
OK for now, but would you please merge the same code in if/else block later to reduce repetition later?
Thanks

Reply:
Yep, will do